### PR TITLE
Fix S2583/S2589 FP: Not tracked collection types can learn `CollectionConstraint` from `BinaryOperation`

### PR DESCRIPTION
**Pull request by Tim-Pohlmann**
Fixes and issue introduced in #7877 
```
class NotACollection
{
    List<int> items;
    int Count => items.Count;

    void Remove(int i) => items.Remove(i);

    void Foo(NotACollection notACollection)
    {
        if (notACollection.Count > 0)
        {
            notACollection.Remove(5);
            if (notACollection.Count > 0)   // Noncompliant FP
            {
                Console.WriteLine();
            }
        }
    }
}
```

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/CollectionConstraintTests.cs [hunk 1]
```diff
@@ -97,4 +97,31 @@ void Test(List<string> foo)
...
+                int Count => items.Count;
+            
+                void Remove(int i) => items.Remove(i);
+            
+                void Foo(NotACollection notACollection)
+                {
+                    if (notACollection.Count > 0)
+                    {
+                        notACollection.Remove(5);
+                        if (notACollection.Count > 0)   // Compliant
```
Comment (mary-georgiou-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/CollectionConstraintTests.cs:118]: Do we have a test where the collection is custom and derives from a tracked collection?
To cover the `.DerivesOrImplementsAny(TrackedCollectionTypes))` case.

Comment (Tim-Pohlmann) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/CollectionConstraintTests.cs:118]: We have one where a custom collection is implementing `IList` but I don't think we have one for inheritance. I can add one next week.

Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/CollectionConstraintTests.cs:118]: Sounds good!

**Review by mary-georgiou-sonarsource — CHANGES_REQUESTED** [review]
LGTM generally.
One missing test :)

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**